src/strings/ops.c

```diff
@@ -522,36 +545,93 @@ MVMString * MVM_string_concatenate(MVMThreadContext *tc, MVMString *a, MVMString
             }
+            if (renormalized_section) {
+                int index_ss_re;
+                int index_ss_a = strands_a - 1;
+                /* Tweak the end of the last strand of string a. Since if a is made up of multiple strands, we can't just refer to index 0 and instead erfer to strands_a - 1 */
+                MVMStringStrand *ss_a = &(result->body.storage.strands[index_ss_a]);
+                MVMStringStrand *ss_re = NULL;
+                ss_a->end -= consumed_a;
+                /* If the strands ends up to be zero length we need to reduce the number of strand_index and also incease lost_strands so the next operation writes over it */
+                if (ss_a->start == ss_a->end)
+                    lost_strands++;
+            /* END 1 */
+            /* START 1.5 (only triggered in some cases) */
+                index_ss_re = strands_a - lost_strands;
+                ss_re = &(result->body.storage.strands[index_ss_re]);
+
+                /* Add the renormalized section in as a strand */
+                ss_re->blob_string = renormalized_section;
+                ss_re->start       = 0;
+                ss_re->end         = renormalized_section->body.num_graphs;
+                ss_re->repetitions = 0;
+                if (ss_re->start == ss_re->end) {
+                    MVM_gc_root_temp_pop(tc);
+                    MVM_exception_throw_adhoc(tc, "Unexpected error in concatenation: renormalized_section is 0 graphemes.\n");
+                    lost_strands++;
+                }
+            /* END 1.5 */
+            }
+            /* START 2 */
+            index_ss_b = strands_a - lost_strands + (renormalized_section_graphs ? 1 : 0 );
             if (effective_b->body.storage_type == MVM_STRING_STRAND) {
-                copy_strands(tc, effective_b, 0, result, strands_a, strands_b);
+                copy_strands(tc, effective_b, 0, result, index_ss_b, strands_b);
             }
             else {
-                MVMStringStrand *ss = &(result->body.storage.strands[strands_a]);
-                ss->blob_string = effective_b;
-                ss->start       = 0;
-                ss->end         = effective_b->body.num_graphs;
-                ss->repetitions = 0;
+                MVMStringStrand *ss_b = &(result->body.storage.strands[index_ss_b]);
+                ss_b->blob_string = effective_b;
+                ss_b->start       = 0;
+                ss_b->end         = effective_b->body.num_graphs;
+                ss_b->repetitions = 0;
+            }
+            if (renormalized_section_graphs) {
+                /* Tweak the beginning of the first strand of string b */
+                MVMStringStrand *ss_b = &(result->body.storage.strands[index_ss_b]);
+                ss_b->start += consumed_b;
+                if (ss_b->start == ss_b->end) {
+                    lost_strands++;
+                    copy_strands(tc, result, index_ss_b + 1, result, index_ss_b, strands_b - 1);
```

Comment:
At the moment `copy_strands` uses `memcpy` rather than `memmove`, and that can go badly wrong when [the source and destination overlap](http://man7.org/linux/man-pages/man3/memcpy.3.html).